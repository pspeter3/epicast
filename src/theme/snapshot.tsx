import * as React from "react";
import TestRenderer from "react-test-renderer";

export const snapshot = (name: string, SFC: React.SFC<{}>): void => {
    describe(name, () => {
        it(`should match snapshot`, () => {
            const component = TestRenderer.create(<SFC>Snapshot</SFC>);
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
};
