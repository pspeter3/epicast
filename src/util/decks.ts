import { Deck } from "../core/types";
import { size } from "./stacks";

export const unique = (deck: Deck): string[] =>
    Object.keys(
        deck.reduce(
            (set, stack) => {
                return Object.keys(stack).reduce((collection, key) => {
                    collection[key] = null;
                    return collection;
                }, set);
            },
            {} as Record<string, null>,
        ),
    );

export const total = (deck: Deck): number => deck.reduce((sum, stack) => sum + size(stack), 0);
