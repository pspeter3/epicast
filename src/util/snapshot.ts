import { shallow } from "enzyme";
import * as React from "react";

export type Factory = () => React.ReactElement<any>;

export const snapshot = (factory: Factory): (() => void) => {
    return () => {
        expect(shallow(factory())).toMatchSnapshot();
    };
};

export const snapshotSuite = (name: string, factory: Factory): void => {
    describe(name, () => {
        it("should match snapshot", snapshot(factory));
    });
};
