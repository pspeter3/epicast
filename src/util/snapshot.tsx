import * as React from "react";
import TestRenderer from "react-test-renderer";

export type Factory = () => React.ReactElement<any>;

export const snapshot = (factory: Factory): (() => void) => {
    return () => {
        const component = TestRenderer.create(factory());
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    };
};

export const snapshotSuite = (name: string, factory: Factory): void => {
    describe(name, () => {
        it("should match snapshot", snapshot(factory));
    });
};
