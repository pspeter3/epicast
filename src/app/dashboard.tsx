import * as React from "react";
import { gameForecast } from "../core/selectors";
import { Game } from "../core/types";
import { ActionProps, Appbar } from "../theme/appbar";
import { Banner } from "../theme/banner";
import { DataTable } from "../theme/data_table";
import { AlertIcon, BuoyIcon, CrosshairIcon, EditIcon, UndoIcon, XIcon } from "../theme/icons";
import { Level } from "../theme/level";
import { Picker } from "../theme/picker";
import { Routes } from "./routes";

export interface Props {
    game: Game;
    onEpidemic: (city: string) => void;
    onRemove: (city: string) => void;
    onUndo: () => void;
}

type Headers = "name" | "infections" | "epidemics";

export class Dashboard extends React.PureComponent<Props, {}> {
    public static displayName = "Dashboard";

    private _action: ActionProps = {
        icon: BuoyIcon,
        href: Routes.Dashboard,
    };

    private _actions: ActionProps[] = [
        {
            icon: UndoIcon,
            href: Routes.Dashboard,
            onClick: evt => {
                evt.preventDefault();
                this.props.onUndo();
            },
        },
        { icon: AlertIcon, href: Routes.Infect },
        { icon: EditIcon, href: Routes.Settings },
    ];

    public render() {
        const { game, onEpidemic, onRemove } = this.props;
        const forecast = gameForecast(game);
        const discards = Object.keys(game.discard).sort();
        const epidemics = Object.keys(game.infection[0]).sort();
        const isSafe = forecast.safe !== 0;
        return (
            <>
                <Appbar action={this._action} title="Epidemia" actions={this._actions} />
                <Level
                    tiles={[
                        { caption: "Turn", value: game.turns + 1 },
                        { caption: "Player Deck", value: forecast.remaining },
                        { caption: "Epidemic", value: forecast.epidemics, isPercent: true },
                    ]}
                />
                {isSafe ? <Banner>No epidemics for at least {forecast.safe} turns</Banner> : null}
                <DataTable<Headers>
                    headers={{ name: false, infections: true, epidemics: true }}
                    defaultSort="infections"
                    data={forecast.cities as Array<Record<Headers, string | number>>}
                />
                <Picker
                    icon={CrosshairIcon}
                    label="Epidemic"
                    options={epidemics}
                    disabled={isSafe}
                    onChange={onEpidemic}
                />
                <Picker
                    icon={XIcon}
                    label="Remove"
                    options={discards}
                    disabled={discards.length === 0}
                    onChange={onRemove}
                />
            </>
        );
    }
}
