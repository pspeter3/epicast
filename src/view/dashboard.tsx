import * as React from "react";
import { Meter, Percentage } from "../theme/data";
import { Main, Row, Section } from "../theme/layout";
import { Header, Text } from "../theme/typography";

export const Dashboard: React.SFC<{}> = () => {
    return (
        <React.Fragment>
            <Section>
                <Row>
                    <Header>Epidemic Probability</Header>
                    <Percentage value={0.63} />
                </Row>
                <Row>
                    <Meter active value={0.63} />
                </Row>
            </Section>
            <Main>
                <Row>
                    <Text>Table Goes Here</Text>
                </Row>
            </Main>
        </React.Fragment>
    );
};
