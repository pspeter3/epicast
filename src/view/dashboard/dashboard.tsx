import * as React from "react";
import { gameForecast } from "../../core/selectors";
import { CityForecast, Forecast, Game } from "../../core/types";
import { Meter, Percentage } from "../../theme/data";
import { Row, Section } from "../../theme/layout";
import { Header } from "../../theme/typography";

export interface Props {
    game: Game;
}

export interface State {
    forecast: Forecast;
    sort: keyof CityForecast;
}

export class Dashboard extends React.PureComponent<Props, State> {
    public static getDerivedStateFromProps(nextProps: Props): Partial<State> {
        return { forecast: gameForecast(nextProps.game) };
    }

    public state: State = {
        forecast: { epidemics: 0, cities: [] },
        sort: "name",
    };

    public render() {
        const { forecast } = this.state;
        const { epidemics } = forecast;
        return (
            <React.Fragment>
                <Section>
                    <Row>
                        <Header>Epidemic Probability</Header>
                        <Percentage value={epidemics} />
                    </Row>
                    <Row>
                        <Meter value={epidemics} />
                    </Row>
                </Section>
            </React.Fragment>
        );
    }
}
