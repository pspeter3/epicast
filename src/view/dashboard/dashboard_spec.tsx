import * as React from "react";
import { snapshotSuite } from "../../util/snapshot";
import { Dashboard } from "./dashboard";

snapshotSuite("Dashboard", () => (
    <Dashboard game={{ player: [8], turns: 0, epidemics: 0, discard: {}, infection: [] }} />
));
