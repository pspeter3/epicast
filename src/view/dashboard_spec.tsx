import * as React from "react";
import { snapshotSuite } from "../util/snapshot";
import { Dashboard } from "./dashboard";

snapshotSuite("Dashboard", () => <Dashboard />);
