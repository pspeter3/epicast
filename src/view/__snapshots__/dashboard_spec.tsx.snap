// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should match snapshot 1`] = `
Array [
  <section
    className="section"
  >
    <div
      className="row"
    >
      <h2
        className="header"
      >
        Epidemic Probability
      </h2>
      <p
        className="data-percentage"
      >
        63
        %
      </p>
    </div>
    <div
      className="row"
    >
      <meter
        className="data-meter active"
        value={0.63}
      />
    </div>
  </section>,
  <main
    className="main"
  >
    <div
      className="row"
    >
      <p
        className="text"
      >
        Table Goes Here
      </p>
    </div>
  </main>,
]
`;
